### PR TITLE
HTTP/1 chunk encoder & task scheduling revamp

### PR DESCRIPTION
**Pull request by graebm**
This PR originally just revamped how HTTP/1 chunks are encoded.
A separate PR https://github.com/awslabs/aws-c-http/pull/274 revamped how HTTP/1 task-scheduling works. It was merged into this branch, on its way to `master`. This fixes issue https://github.com/awslabs/aws-c-http/issues/270

Encoder changes:
- Combine the 2 state machines into 1.
- Encoder no longer deals with the lock when getting new chunks. The stream will be responsible for getting new chunks into the thread.

Task-Scheduling changes:
- HTTP/1 connection shares 1 lock with all its streams.
- connection has new `cross_thread_work_task` that moves work from `synced_data` to `thread_data`.
  - kicks off the `outgoing_stream_task` if necessary
- similarly, stream has a `cross_thread_work_task` for moving work from `synced_data` to `thread_data`.
  - it can also kick off the connection's `outgoing_stream_task` if necessary

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: include/aws/http/private/h1_encoder.h [hunk 1]
```diff
@@ -5,37 +5,19 @@
...
-    struct aws_http1_stream_chunk *current_chunk;
-    bool paused;
-};
-
-enum aws_h1_encoder_body_stream_state {
-    AWS_H1_ENCODER_STATE_CHUNK_INIT,
-    AWS_H1_ENCODER_STATE_CHUNK_LINE,
-    AWS_H1_ENCODER_STATE_CHUNK_PAYLOAD,
-    AWS_H1_ENCODER_STATE_CHUNK_END,
-    AWS_H1_ENCODER_STATE_CHUNK_TERMINATED,
```
Comment (graebm) [inline include/aws/http/private/h1_encoder.h:35]: Removing this 2nd state-machine enum and combining into a single state-machine enum

File: source/h1_encoder.c [hunk 2]
```diff
@@ -351,110 +356,27 @@ int aws_h1_encoder_start_message(
...
     encoder->message = message;
-    encoder->state = AWS_H1_ENCODER_STATE_HEAD;
-    encoder->progress_bytes = 0;
-
-    return AWS_OP_SUCCESS;
-}
-
-static int s_h1_encoder_process_content_length_body(struct aws_byte_buf *dst, struct aws_h1_encoder *encoder) {
-    while (dst->capacity >= dst->len) {
-        if (dst->capacity == dst->len) {
```
Comment (graebm) [inline source/h1_encoder.c:365]: Github was doing such a terrible job showing the diffs on this function I just moved it below to a better place.

File: source/h1_encoder.c [hunk 3]
```diff
@@ -712,116 +443,308 @@ int aws_chunk_line_from_options(struct aws_http1_chunk_options *options, struct
...
-        }
-        if (AWS_UNLIKELY(AWS_OP_SUCCESS != s_transfer_encode_chunk(dst, encoder))) {
+/* Write as much body stream as possible into dst buffer.
+ * Increments encoder->progress_bytes to track progress */
+static int s_encode_stream(
+    struct aws_h1_encoder *encoder,
+    struct aws_byte_buf *dst,
+    struct aws_input_stream *stream,
+    uint64_t total_length,
+    bool *out_done) {
```
Comment (graebm) [inline source/h1_encoder.c:468]: This is that giant function I moved

File: source/h1_encoder.c [hunk 4]
```diff
@@ -712,116 +443,308 @@ int aws_chunk_line_from_options(struct aws_http1_chunk_options *options, struct
...
     return AWS_OP_SUCCESS;
 }
 
-static int s_h1_encoder_process_body(struct aws_byte_buf *dst, struct aws_h1_encoder *encoder) {
+/* A state function should:
+ * - Raise an error only if unrecoverable error occurs.
+ * - `return s_switch_state(...)` to switch states.
+ * - `return AWS_OP_SUCCESS` if it can't progress any further (waiting for more
+ *    space to write into, waiting for more chunks, etc). */
+typedef int encoder_state_fn(struct aws_h1_encoder *encoder, struct aws_byte_buf *dst);
```
Comment (graebm) [inline source/h1_encoder.c:550]: From here on down is the new state machine. The logic of how bytes are copied is basically the same as before. But now each state is its own function, and there's one super simple state machine loop that drives (loop is at bottom of file)

File: source/h1_encoder.c [hunk 5]
```diff
@@ -712,116 +443,308 @@ int aws_chunk_line_from_options(struct aws_http1_chunk_options *options, struct
...
-            "Writing to message, outgoing head progress %" PRIu64 "/%zu.",
-            encoder->progress_bytes,
-            encoder->message->outgoing_head_buf.len);
+    return s_switch_state(encoder, AWS_H1_ENCODER_STATE_CHUNK_BODY);
+}
+
+/* Write out data for current chunk */
+static int s_state_fn_chunk_body(struct aws_h1_encoder *encoder, struct aws_byte_buf *dst) {
+    bool done;
+    if (s_encode_stream(encoder, dst, encoder->message->body, encoder->message->content_length, &done)) {
```
Comment (TingDaoK) [inline source/h1_encoder.c:664]: I haven't seen where we update the `encoder->message->content_length` to `encoder->message->body_chunks->current_chunk->data_size`. We may need to update it? Or should it be `s_encode_stream(encoder, dst, encoder->message->body, encoder->message->body_chunks->current_chunk->data_size, &done)` ?

Comment (TingDaoK) [reply source/h1_encoder.c:664]: Oh, the `encoder->message->body` should be `encoder->current_chunk->data`?

File: source/h1_encoder.c [hunk 6]
```diff
@@ -712,116 +443,308 @@ int aws_chunk_line_from_options(struct aws_http1_chunk_options *options, struct
...
+    ENCODER_LOGF(
+        TRACE,
+        encoder,
+        "Sending %zu bytes of body, progress: %" PRIu64 "/%" PRIu64,
+        amount_read,
+        encoder->progress_bytes,
+        total_length);
+
+    /* Return if we're done sending stream */
+    if (encoder->progress_bytes == encoder->message->content_length) {
```
[the remark this answers is not in the document]
Comment (graebm) [reply source/h1_encoder.c:510]: great catch, this is totally wrong 🤦‍♂️

File: source/h1_encoder.c (hunk 5 above)
Comment (graebm) [reply source/h1_encoder.c:664]: yup 🤦‍♂️

**Pull request merged**